src/rdb/rdb_tx.c

```diff
@@ -417,20 +420,58 @@ rdb_tx_commit(struct rdb_tx *tx)
 	 */
 	if (!rdb_tx_is_critical(tx)) {
 		daos_size_t	scm_remaining = 0;
+		uint32_t	nchecks = 0;
 
+check_space:
 		rc = rdb_scm_left(tx->dt_db, &scm_remaining);
 		if (rc != 0) {
 			D_ERROR(DF_DB": failed to query free space\n",
 				DP_DB(tx->dt_db));
 			goto out_lock;
 		}
+		nchecks++;
 
 		if (scm_remaining < RDB_NOAPPEND_FREE_SPACE) {
-			D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not "
-			       "append! scm_left="DF_U64"\n", DP_DB(tx->dt_db),
-			       scm_remaining);
-			D_GOTO(out_lock, rc = -DER_NOSPACE);
+			uint64_t		idx = 0;
+			uint64_t		delta_nsec;
+
+			if (nchecks > 1) {
+				D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not append! "
+				       "scm_left="DF_U64"\n", DP_DB(tx->dt_db), scm_remaining);
+				D_GOTO(out_lock, rc = -DER_NOSPACE);
+			} else {
+				/* Throttle number of times compacting the log when low on space. */
+				delta_nsec = (d_hlc2nsec(d_hlc_get()) -
+					      d_hlc2nsec(tx->dt_db->d_nospc_ts));
+				if (delta_nsec < RDB_NOSPC_ERROR_INTVL_NSEC) {
+					D_DEBUG(DB_TRACE, DF_DB": nearly out of space, but too "
+						"early to trigger compaction\n", DP_DB(tx->dt_db));
+					goto check_space;
```

Comment:
[Question] Is this a busy polling loop? Should we yield somehow?

Reply:
I see. Sorry for the noise. This makes me wondering, if there's a way to achieve a complete GC round inside vos, then when dlr_aggregated == base (or should it be base - 1? I forgot), we know there is any space to be reclaim and can return -DER_NOSPACE directly.

Reply:
By "achieving a complete GC round", I mean "triggering and waiting for completion", similar to what we do for dlr_aggregate in this patch.

Reply:
Thanks. I can study this potential and consider a future follow-on patch.